Feature/pw/gradebook get staff

pwilkins opened the pull request:
Return staff membership for a gradebook. The data returned is de-normalized from what is returned by the LMod Web service.  Due to the comprehensive amount of data returned, a 'simple' parameter is available to return less data.
This PR includes the implementation of GradeBook.get_options() since it may be a dependency of get_staff()

Not ready for review.  @carsongee @pdpinch @Ferdi

carsongee commented on pylmod/gradebook.py:
```diff
@@ -101,6 +100,64 @@ def get_gradebook_id(self, gbuuid):
+        end_point = 'gradebook/options/{gradebookId}'.format(
+            gradebookId=gradebook_id or self.gradebook_id)
+        options = self.get(end_point)
+        return options
```
you'll want to return `options['data']`as we do elsewhere

carsongee commented on pylmod/tests/test_gradebook.py:
```diff
@@ -270,6 +317,20 @@ def test_get_gradebook_id(self):
+        gradebook = GradeBook(self.CERT, self.URLBASE, self.GBUUID)
+        options = gradebook.get_options(gradebook_id=self.GRADEBOOK_ID)
+        self.assertEqual(
+            options['data']['membershipQualifier'],
```
as above, you shouldn't have to ref `data`

pwilkins commented:
This PR is ready for review.  @carsongee @pdpinch

pdpinch commented on pylmod/gradebook.py:
```diff
@@ -76,6 +76,25 @@ def unravel_sections(section_data):
         return sections
 
+    @staticmethod
+    def unravel_staff(staff_data):
```
_unravel_staff() perhaps?

Is this method limited to unraveling staff data?

pwilkins replied:
If we were to change this to _unravel_staff(), we'd also want to change unravel_sections() to _unravel_sections().

And, "Yes", it is limited to unraveling staff.

carsongee replied:
It isn't private because we don't unravel by default.  Without `simple=True` we return the structure as returned by the service and want to provide a way to unravel that data and keep the full "non-simple" structure to consumers of the class.

pdpinch commented on pylmod/gradebook.py:
```diff
@@ -948,3 +1024,152 @@ def spreadsheet2gradebook(
+
+        Get staff list for the gradebook specified.
+
+        If simple=True, return a list of dictionaries, one dictionary
```
Do we want an example of the return when simple=True -- perhaps this description is sufficient?

pwilkins replied:
@pdpinch @carsongee An example would be good idea, but a better question would be, "Which of these member attributes belong in a 'simple' list?".  I'm currently returning all of them but Shirley that is more than anyone needs.

carsongee replied:
https://github.com/mitodl/PyLmod/blob/feature/pw/gradebook_get_staff/pylmod/gradebook.py#L806-L812

pwilkins replied:
I misspoke.  I'm currently returning just displayName, accountEmail, and role for the "simple" list.

pdpinch replied:
Should we add personId and call it a day? (I'm assuming personId is a parameter for other API calls, the same way the API uses courseid and gradebookid)

pdpinch replied:
Ok. Carson's persuaded me that we don't need personId -- and we can add it later if we do.

carsongee commented on pylmod/gradebook.py:
```diff
@@ -948,3 +1024,152 @@ def spreadsheet2gradebook(
+                    'role': member['role'],
+                })
+            return simple_list
+        return staff_data
```
shouldn't this be staff_data['data'] like the others, or does it not have that key?

carsongee commented on pylmod/tests/test_gradebook.py:
```diff
@@ -110,6 +110,195 @@ class TestGradebook(BaseTest):
+        u'status': 1
+    }
+
+    SIMPLE_STAFF_BODY = [
```
You should be able to independently generate this to save the space in the file

carsongee commented on pylmod/tests/test_gradebook.py:
```diff
@@ -110,6 +110,195 @@ class TestGradebook(BaseTest):
+        },
+    }
+
+    STAFF_BODY = {
```
Wouldn't this be adequate as just two users in two different roles?  I don't see a test for checking that each role type is in the return data.  Generally, I'd prefer to only include test data that is actually tested.  That said, if you want to add tests so that all of this data is used, that would be ideal.

carsongee commented on pylmod/tests/test_gradebook.py:
```diff
@@ -110,6 +110,195 @@ class TestGradebook(BaseTest):
+            ]
+        },
+        u'message': u'',
+        u'status': 1
```
do we use these fields in any of our code?

carsongee replied:
I should be more specific, are `message` or `status` ever accessed in our code?

pwilkins replied:
We never use these fields.  I'll pare down the returned data.

carsongee commented on pylmod/tests/test_gradebook.py:
```diff
@@ -110,6 +110,195 @@ class TestGradebook(BaseTest):
         ]
     }
 
+    OPTIONS_BODY = {
```
Same with this data structure, if we don't have code that is validating any of this data, other than it's equivalency, you can use the simple structs and just create it in the test.  Also, if you are only using it in one test, it should be declared in that test, so this should just be something like:
```python
  @httpretty.activate
    def test_get_options(self):
        """Verify that we can get the options for a gradebook."""
        test_body = {'data': { 'membershipQualifier': '/project/mitxdemosite'' }}
        httpretty.register_uri(
            httpretty.GET,
            '{0}gradebook/options/{1}'.format(
                self.GRADEBOOK_REGISTER_BASE,
                self.GRADEBOOK_ID
            ),
            body=json.dumps(self.test_body)
        )
        gradebook = GradeBook(self.CERT, self.URLBASE, self.GBUUID)
        options = gradebook.get_options(gradebook_id=self.GRADEBOOK_ID)
        self.assertEqual(
            options['membershipQualifier'],
            test_body['data']['membershipQualifier']
        )
```
instead of having a class level static structure and URL register.

carsongee commented on pylmod/tests/test_gradebook.py:
```diff
@@ -270,6 +480,20 @@ def test_get_gradebook_id(self):
+            options['membershipQualifier'],
+            self.OPTIONS_BODY['data']['membershipQualifier']
+        )
+        # check for no data
```
There isn't actually a check for not data here, so pull this or actually test it.

pwilkins commented:
@carsongee @pdpinch ready for review.

pdpinch commented:
I am satisfied, but coveralls is not. Looks like there's a branch in the test code that's never reached:  https://coveralls.io/builds/2335678/source?filename=pylmod%2Ftests%2Ftest_gradebook.py#L245

carsongee commented on pylmod/tests/test_gradebook.py:
```diff
@@ -267,7 +345,31 @@ def test_get_gradebook_id(self):
+            test_body['data']['membershipQualifier']
+        )
+        # check for no data
+        self._register_get_options(False)
```
It appears you are now doing it both ways, you are now using that URL twice, but you are registering for it in two different locations.  Additionally, you still don't need a method for this, just delete the `data` key from `test_body` and you won't get data back, no need for branches and a function.

carsongee commented on pylmod/tests/test_gradebook.py:
```diff
@@ -437,17 +539,39 @@ def test_get_sections(self):
         )
         self.assertEqual(
-            assignments,
+            sections,
```
whoops :blush:

carsongee commented on pylmod/tests/test_gradebook.py:
```diff
@@ -110,6 +110,59 @@ class TestGradebook(BaseTest):
+                    u'email': u'benfranklin@mit.edu',
+                    u'givenName': u'Benjamin',
+                    u'middleName': None,
+                    u'mitId': u'921344431',
```
I would strip out all the properties but the ones returned by simple here.

pwilkins commented:
@pdpinch @carsongee This is again ready for review.

carsongee commented on pylmod/tests/test_gradebook.py:
```diff
@@ -437,17 +511,39 @@ def test_get_sections(self):
+        """Verify staff list is returned."""
+        self._register_get_gradebook()
+        gradebook = GradeBook(self.CERT, self.URLBASE, self.GBUUID)
+        self._register_get_staff()
```
This is only called once, so go ahead and inline it.

carsongee commented on pylmod/tests/test_gradebook.py:
```diff
@@ -267,7 +327,21 @@ def test_get_gradebook_id(self):
+        self.assertIn('membershipQualifier', options)
+
+        # check for no data
+        self._register_get_options(False)
```
To my earlier comment, you could have just reused the URI registration and changed the returned data with `del test_data['data']` and avoided needing a function or `if..then`, but this is fine as is since you already solved it this way.

carsongee replied:
This one is my bad as discussed IRL. You may want to change it to a function that accepts the body though so that you don't need conditional statements, but it is fine as is

carsongee commented:
Ready to merge after that last little nit pick, or if you just want me to merge it as is, I'm fine with that too.

carsongee commented:
Oh, and you should at least add to the commit message that we added get_options since we ended up munging two features into one.

pwilkins commented:
I have made what I hope are the final edits to address review comments.  I squashed the last two commits and updated the commit message to include Gradebook.get_options().  This should be ready to merge, unless there is more squashing needed.  @pdpinch @carsongee

carsongee commented on pylmod/gradebook.py:
```diff
@@ -76,6 +76,25 @@ def unravel_sections(section_data):
 
+    @staticmethod
+    def unravel_staff(staff_data):
+        """Unravels section type dictionary into flat list of sections with
```
s/section/staff

carsongee commented:
Other than that last little typo, :helicopter:

pwilkins commented:
fixed the typo(s) in unravel_staff() docstring.  (Good catch!)  Ready for merge.  @carsongee @pdpinch
